File: solutions/3144. Minimum Substring Partition of Equal Character Frequency/3144.py
# Time complexity: O(n^2)
# Space complexity: O(n)
class Solution:
  def minimumSubstringsInPartition(self, s: str) -> int:
    n = len(s)
    dp = [0] * (n + 1)
    dp[0] = 0

    for i in range(1, n + 1):
      dp[i] = i
      count = {}
      for j in range(i, 0, -1):
        count[s[j - 1]] = count.get(s[j - 1], 0) + 1
        if len(set(count.values())) == 1:
          dp[i] = min(dp[i], 1 + dp[j - 1])

    return dp[n]

File: solutions/3144. Minimum Substring Partition of Equal Character Frequency/3144.cpp
// Time complexity: O(n^2)
// Space complexity: O(n)
class Solution {
 public:
  int minimumSubstringsInPartition(string_view s) {
    const int n = s.length();
    vector<int> dp(n, n);

    for (int i = 0; i < n; ++i) {
      vector<int> count(26);
      for (int j = i; j >= 0; --j) {
        ++count[s[j] - 'a'];
        if (isBalanced(count))
          dp[i] = j > 0 ? min(dp[i], 1 + dp[j - 1]) : 1;
      }
    }

    return dp.back();
  }
 private:
  static constexpr int kMax = 1001;

  bool isBalanced(const vector<int>& count) {
    int minFreq = kMax;
    int maxFreq = 0;
    for (const int freq : count)
      if (freq > 0) {
        minFreq = min(minFreq, freq);
        maxFreq = max(maxFreq, freq);
      }

    return minFreq == maxFreq;
  }
};